File: leetcode/medium/113-path-sum-2/rec-btrack.cpp
/**
 * @file rec-btrack.cpp
 * @author Vortexx2
 * @brief Problem 113 - Path Sum II
 * @date 29-07-2021
 * 
 * Runtime - 4 ms
 * Memory Usage - 19.7 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <algorithm>
#include <list>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  vector<vector<int>> pathSum(TreeNode *root, int targetSum)
  {
    vector<vector<int>> res;
    vector<int> path;

    helper(root, targetSum, path, res);
    return res;
  }

  void helper(TreeNode *curr, int targetSum, vector<int>& path, vector<vector<int>>& res)
  {
    if (!curr)
      return;
    path.push_back(curr->val);

    if (!curr->left && !curr->right && curr->val == targetSum)
    {
      res.push_back(path);
    }

    helper(curr->left, targetSum - curr->val, path, res);
    helper(curr->right, targetSum - curr->val, path, res);

    path.pop_back();
  }
};

File: leetcode/medium/113-path-sum-2/rec-btrack.py
"""
  @author Vortexx2
  Problem 113 - Path Sum II

  Runtime - 44 ms
  Memory Usage - 15.3 MB
"""
from typing import List


class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def pathSum(self, root: TreeNode, targetSum: int) -> List[List[int]]:

    res, path = [], []
    self.helper(root, targetSum, path, res)
    return res

  def helper(self, curr: TreeNode, sm: int, path: List[int], res: List[List[int]]) -> None:

    if not curr:
      return

    path.append(curr.val)

    if not curr.left and not curr.right and curr.val == sm:
      res.append(path.copy())

    self.helper(curr.left, sm - curr.val, path, res)
    self.helper(curr.right, sm - curr.val, path, res)

    path.pop()
